File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> ans = new ArrayList<>();
        Map<String, Integer> map = new HashMap<>();
        for(String s:strs){
            char[] ch = s.toCharArray();
            Arrays.sort(ch);
            String a = new String(ch);
            if(map.containsKey(a)){
                ans.get(map.get(a)).add(s);
            }else{
                ans.add(new ArrayList<>());
                int n = ans.size()-1;
                ans.get(n).add(s);
                map.put(a,n);
            }
            
        }
        return ans;
    }
}

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        cache = {}
        res = []
        count = 0
        for word in strs:
            index = ''.join(sorted(word))
            if index in cache:
                res[cache[index]].append(word)
            else:
                res.append([word])
                cache[index] = count
                count += 1
        return res
